### ISSUE TITLE
5.2.6-pre-release not working with older devices

### ISSUE THREAD

**Issue by cdruan** [Jermolene/TiddlyWiki5#7350]
### Describe the bug

Offenders:

* PR #7297 (sticky flag is an ES6 (ES2015) feature)
* `core/modules/filters/strings.js` line 110: `const`

Reverting the \#7297 commit and changing `const` to `var` confirmed to resolve the issue.

Perhaps PR #7297 could be delayed until 5.3.0, so the new markdown plugin has a chance to run on both new and old devices?

### TiddlyWiki Configuration

- Version 5.2.6-pre-release

- Device: iPad 3
- OS: iOS 9.3.5
- Browser: safari

**Comment by pmario** [comment]
> core/modules/filters/strings.js line 110: const

https://github.com/Jermolene/TiddlyWiki5/blob/589813f23335bbd3b50be52ad5d5211fbbffee33/core/modules/filters/strings.js#L110

@Jermolene ... There is a `const` which should probably not be there.

**Comment by pmario** [comment]
@cdruan The PR you mentioned is _not_ responsible for the const definition. I think it's an oversight

**Comment by Jermolene** [comment]
Hi @cdruan I've removed the accidental "const" in f247686970f246fda69925e1f7b663c3dedc6f39.

As you note, the regexp sticky flag [was only introduced in iOS 10](https://caniuse.com/mdn-javascript_builtins_regexp_sticky), so it is still going to be a problem with your iOS 9.3.5 device. I'm inclined to revert #7297 for the moment, as you suggest, until we formally move to a more modern JS subset.

**Comment by pmario** [comment]
>I'm inclined to revert https://github.com/Jermolene/TiddlyWiki5/pull/7297 for the moment, as you suggest, until we formally move to a more modern JS subset.

@linonetwo 
Your thoughts. Can you provide some performance numbers?

**Comment by pmario** [comment]
@Jermolene ... Would it be possible to move the regexp definitions out of the functions, so we can easily patch them and only them with a plugin?

**Comment by pmario** [comment]
Just to be sure. I was thinking about to keep the improvement and add a possibility of old devices to add a plugin for a fix. .. Otherwise we will never move to more modern js versions.

**Comment by Jermolene** [comment]
Hi @pmario 

> Would it be possible to move the regexp definitions out of the functions, so we can easily patch them and only them with a plugin?

That would be fairly reasonable, I think. Some of the parsing regexps from TW Classic are already [pulled out into $tw.config](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/config.js).

> Just to be sure. I was thinking about to keep the improvement and add a possibility for old devices to add a plugin for a fix .. Otherwise we will never move to more modern js versions.

Actually, I'm increasingly thinking that now is the time to move our JS baseline. I'll make another issue about it.

**Comment by saqimtiaz** [comment]
> Actually, I'm increasingly thinking that now is the time to move our JS baseline. I'll make another issue about it.

Yes please. If not for this release, then perhaps for the parameterized transclusions one?

**Comment by pmario** [comment]
>Yes please. If not for this release, then perhaps for the parameterized transclusions one?

I would support this. 

----------

I think we should try to make v5.2.6 work with the status quo and have a closer look for v5.3.0

**Comment by pmario** [comment]
@Jermolene .. I thought they could be defined as parseutils.js. IMO It would be easy to overwrite with a little plugin, `module-type: utils`. .. Functions names are only examples. 

```
/*
Look for a macro invocation parameter. Returns null if not found, or {type: "macro-parameter", name:, value:, start:, end:}
*/

exports.getMacroParameterRegExp = function() {
	return /(?:([A-Za-z0-9\-_]+)\s*:)?(?:\s*(?:"""([\s\S]*?)"""|"([^"]*)"|'([^']*)'|\[\[([^\]]*)\]\]|((?:(?:>(?!>))|[^\s>"'])+)))/y;
}

exports.parseMacroParameter = function(source,pos) {
	var node = {
		type: "macro-parameter",
		start: pos
	};
	// Define our regexp
	// Skip whitespace
	pos = $tw.utils.skipWhiteSpace(source,pos);
	// Look for the parameter
	var token = $tw.utils.parseTokenRegExp(source,pos,this.getMacroParameterRegExp());
	if(!token) {
		return null;
	}
...
...
```

What do you think?

**Comment by Jermolene** [comment]
I've made an issue to discuss updating our target JS dialect at #7353

**Comment by Jermolene** [comment]
> @Jermolene .. I thought they could be defined as parseutils.js. IMO It would be easy to overwrite with a little plugin, `module-type: utils`. .. Functions names are only examples.

I don't think it's worth indirecting through a function to get the regexp, particularly an implementation like that that means that a new regexp object would be instantiated for each invocation.

We already have `module-type: config` whose exports are merged into `$tw.config`:

https://github.com/Jermolene/TiddlyWiki5/blob/f247686970f246fda69925e1f7b663c3dedc6f39/core/modules/startup/load-modules.js#L29

**Comment by cdruan** [comment]
Fixed by commits 308e207 and f247686.

**Comment by linonetwo** [comment]
> Your thoughts. Can you provide some performance numbers?

It seems to be 600ms -> 0.13 ms, roughly, regarding to my screenshot in https://github.com/Jermolene/TiddlyWiki5/issues/7277 and https://github.com/Jermolene/TiddlyWiki5/pull/7297#issuecomment-1446381841
